StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/SavedPaymentMethodManagerTest.swift

```diff
@@ -0,0 +1,177 @@
+    }
+
+    func testDetachPaymentMethod_withLegacyCustomer() {
+        var configuration = configuration
+        configuration.customer = .init(id: "cus_test123", ephemeralKeySecret: ephemeralKey)
+
+        let expectation = stubDetachPaymentMethod(paymentMethod: STPPaymentMethod.stubbedPaymentMethod(),
+                                                  ephemeralKey: ephemeralKey)
+
+        let sut = SavedPaymentMethodManager(configuration: configuration)
+        sut.detach(paymentMethod: paymentMethod, using: ephemeralKey)
+
+        wait(for: [expectation], timeout: 5.0)
+    }
+
+    func testDetachPaymentMethod_withCustomerSession() {
+        var configuration = configuration
+        configuration.customer = .init(id: "cus_test123", customerSessionClientSecret: "session_123")
+
+        let listPaymentMethodsExpectation = stubListPaymentMethods(customerId: configuration.customer!.id,
+                                                                   ephemeralKey: ephemeralKey)
+
+        let detachExpectation = stubDetachPaymentMethod(paymentMethod: STPPaymentMethod.stubbedPaymentMethod(),
+                                                  ephemeralKey: ephemeralKey)
+
+        let sut = SavedPaymentMethodManager(configuration: configuration)
+        sut.detach(paymentMethod: paymentMethod, using: ephemeralKey)
+
+        wait(for: [listPaymentMethodsExpectation, detachExpectation], timeout: 5.0)
+    }
+}
+
+extension SavedPaymentMethodManagerTests {
+    // MARK: HTTP Stubs
+
+    func stubRequest(urlContains: String,
+                     ephemeralKey: String,
+                     httpMethod: String,
+                     responseObject: Any) -> XCTestExpectation {
+        let exp = expectation(description: "Request \(httpMethod) \(urlContains)")
+
+        stub { urlRequest in
+          return urlRequest.url?.absoluteString.contains(urlContains) ?? false
+            && urlRequest.allHTTPHeaderFields?["Authorization"] == "Bearer \(ephemeralKey)"
+            && urlRequest.httpMethod == httpMethod
+        } response: { _ in
+            DispatchQueue.main.async {
+                exp.fulfill()
+            }
+            return HTTPStubsResponse(jsonObject: responseObject,
+                                     statusCode: 200,
+                                     headers: nil)
+        }
+
+        return exp
+    }
+
+    func stubUpdatePaymentMethod(paymentMethod: STPPaymentMethod, ephemeralKey: String) -> XCTestExpectation {
+        return stubRequest(urlContains: "/payment_methods/\(paymentMethod.stripeId)",
+                           ephemeralKey: ephemeralKey,
+                           httpMethod: "POST",
+                           responseObject: STPPaymentMethod.paymentMethodJson)
+    }
+
+    func stubDetachPaymentMethod(paymentMethod: STPPaymentMethod, ephemeralKey: String) -> XCTestExpectation {
+        return stubRequest(urlContains: "/payment_methods/\(paymentMethod.stripeId)/detach",
```

Comment:
Does this assert that the method was called?

Reply:
Not directly but yes, we return the `XCTestExpectation` which is fulfilled [here](18eadf663dff549cbe31cd38aa433b14643a5bc491b64e40483e0701f85a2c00R99), and each test case waits for the expectations.

Reply:
Ah I see now, perfect thanks.